src/libraries/System.Net.Http/src/System/Net/Http/HttpMetrics.cs

```diff
@@ -0,0 +1,107 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Diagnostics.Metrics;
+
+namespace System.Net.Http
+{
+    internal sealed class HttpMetrics
+    {
+        private readonly Meter _meter;
+        private readonly UpDownCounter<long> _currentRequests;
+        private readonly Histogram<double> _requestsDuration;
+
+        public HttpMetrics(Meter meter)
+        {
+            _meter = meter;
+
+            _currentRequests = _meter.CreateUpDownCounter<long>(
+                "current-requests",
+                description: "Number of outbound HTTP requests that are currently active on the client.");
+
+            _requestsDuration = _meter.CreateHistogram<double>(
+                "request-duration",
+                unit: "s",
+                description: "The duration of outbound HTTP requests.");
+        }
+
+        public void RequestStart(HttpRequestMessage request)
+        {
+            if (_currentRequests.Enabled)
+            {
+                RequestStartCore(request);
+            }
+        }
+
+        public void RequestStartCore(HttpRequestMessage request)
+        {
+#pragma warning disable SA1129 // Do not use default value type constructor
+            var tags = new TagList();
+#pragma warning restore SA1129 // Do not use default value type constructor
+            InitializeCommonTags(ref tags, request);
+
+            _currentRequests.Add(1, tags);
+        }
+
+        public void RequestStop(HttpRequestMessage request, HttpResponseMessage? response, Exception? unhandledException, long startTimestamp, long currentTimestamp)
+        {
+            if (_currentRequests.Enabled || _requestsDuration.Enabled)
+            {
+                RequestStopCore(request, response, unhandledException, startTimestamp, currentTimestamp);
+            }
+        }
+
+        private void RequestStopCore(HttpRequestMessage request, HttpResponseMessage? response, Exception? unhandledException, long startTimestamp, long currentTimestamp)
+        {
+#pragma warning disable SA1129 // Do not use default value type constructor
+            var tags = new TagList();
+#pragma warning restore SA1129 // Do not use default value type constructor
+            InitializeCommonTags(ref tags, request);
+
+            _currentRequests.Add(-1, tags);
+
+            if (response is not null)
+            {
+                tags.Add("status-code", (int)response.StatusCode); // Boxing?
+                tags.Add("protocol", $"HTTP/{response.Version}"); // Hacky
+            }
+            if (unhandledException is not null)
+            {
+                tags.Add("exception-name", unhandledException.GetType().FullName);
+            }
+            if (request.HasTags)
+            {
+                foreach (var customTag in request.MetricsTags)
```

Comment:
Similar to HttpRequest properties not being stable, these tags might also not be stable across the duration of the request. Your EnrichmentHandler test is an example that deliberately changes them.

I think one scenario we need to sort out is the case where the caller to HttpClient is some 3rd party library that isn't specifying tags but the app developer does want to get all their traffic classified. Currently R9 is handling this case by having the app developer create a mapping from outbound URIs -> other metadata, then they register a handler in the HttpClientFactory to execute that mapping and generate the tagged telemetry. If R9 is going to switch over to use these built-in metrics rather than continuing to add their own then they will need some way to dynamically inject those tags. If the injection mechanism will continue to be a handler (similar to your EnrichmentHandler), then perhaps we need to move the metric collection point back to the DiagnosticsHandler where it is now so that other handlers have a chance to update those tags before they are captured. (I know I suggested doing it at the front earlier, I'm not sure my idea was a good one now that I consider this scenario)

@dpk83 - you probably want to take a look at this if you haven't already

Reply:
I don't understand the concern here. R9 has a DelegatingHandler that is responsible for recording to the metric. The handler also executes enrichers to build up a collection of custom tags.

The difference in this prototype is recording to the metric and gathering custom tags happens in different places. The handler writes them onto the `HttpRequestMessage`, and they are read off the message later when needed.

My main concern with this approach is making sure there is the opportunity for an enrichment handler to add tags after `SendAsync` has returned. Right now the metrics are in the `HttpMessageInvoker`. It works, but it doesn't record separate HTTP requests if a Polly handler is added for retries. The solution is to move metrics to `DiaganosticsHandler`. That handler is always inside a retry handler, so each HTTP request is recorded, but we lose the ability to do enrichment after `SendAsync` returns. `DiaganosticsHandler` is added after any user-injected handlers (such as enrichment), so it's problematic to log it there.

A solution that supports both goals could be to wait to record the HTTP duration counter value when an HTTP request has completed both of these actions:
1. The HttpRequestMessage exits the HTTP pipeline
2. The HttpResponseMessage finishes reading the response

It would be slightly odd that `DiaganosticsHandler` "starts" metrics for an HTTP request (e.g. starts the request duration timer, increments current request count), and exiting `HttpMessageInvoker` "finishes" it (e.g. records request duration with enriched tags and decreases current request count).

Reply:
Sorry for confusion, upon re-reading the code I see that I misunderstood it earlier and my comment was based on that misunderstanding. I was incorrectly worried that custom tags would cause the current-requests increment operation to be recorded with different tags than the decrement which would unbalance it. Now I see that custom tags are only used for the duration measurement and my concern was unfounded. 

While I do still care about the scenario I described above with 3rd party libraries, I think this design would be able to handle it.

Reply:
Few points:
- Quite often caller doesn't have access to the HttpRequestMessage, third party library is a very good example of this scenario. In complex organizations where there are hundreds and thousands of services a common libraries are used quite a lot and this option just doesn't work. We devised IDownstreamDependencyMetadata for that very purpose.

- We are now supporting both DelegatingHandler based auto collection of metrics as well as DiagnosticsSource based approach because DelegatingHandler wasn't sufficient for a lot of cases.

- In order to consume this metrics directly emitted from .NET and provide the transformation and the dependency metadata looks like we will still continue to use DelegatingHandler and DiagnosticsSource so I don't see how we can take consume and build onto the metrics this component is going to emit without incurring additional perf overhead.

- Is this support is going to be ported back for .NET Framework?

Reply:
> The solution is to move metrics to `DiaganosticsHandler`

Note that `DiagnosticsHandler` is now strongly coupled with distributed tracing, and disabling distributed tracing either by by setting setting `DOTNET_SYSTEM_NET_HTTP_ENABLEACTIVITYPROPAGATION=0` or by setting `socketsHttpHandler.ActivityHeadersPropagator = null` would completely disable it.

Wouldn't it be better to implement metrics in a separate `MetricsHandler` that would live either before or after `DiagnosticsHandler` in the chain?

Reply:
Sure, if a separate handler is better, then do it. The difference we're considering is the same: counters at the HttpMessageInvoker layer vs counters at an inner handler.

Reply:
> I don't see how we can take consume and build onto the metrics this component is going to emit without incurring additional perf overhead.

What perf overhead are you worried about @dpk83? 

While I don't think R9 strictly has to use this counter for correct behavior, from the consumer's perspective I'd really hope to describe the R9 stuff to 3P devs as an enhancement and not another orthogonal telemetry mechanism. I am leaving out here 1P back-compat where we probably want a different schema. For example I don't think it would be good if 3P devs see duplicate measurements. That would mean that R9 either needs to add tags to this instrument or suppress the instrument and replicate it, including the new enrichment tags. Making sure an alternate implementation stays in-sync feels possible, but error-prone and more work intensive.

Reply:
> Another question, will the same metrics support be added for net framework or is it limited to net core only?

I think this got answered during our meeting earlier, but reproducing it here - the new metrics would only be present in .NET 8 onwards. This code doesn't flow downlevel either to framework or to older versions of .NET Core.